robotoff/prediction/category/neural/keras_category_classifier_3_0/preprocessing.py

```diff
@@ -0,0 +1,272 @@
+    :return: a KeywordProcessor
+    """
+    # dict mapping an normalized ingredient to a set of (node ID, lang) tuples
+    name_map = defaultdict(set)
+    # it's the reverse structure as name_map, dict mapping a (node ID, lang)
+    # to a set of normalized names. Used to generate synonym combinations
+    synonyms = defaultdict(set)
+    for node in ingredient_taxonomy.iter_nodes():
+        if node.id in INGREDIENT_ID_EXCLUDE_LIST:
+            # Ignore ingredients that create false positive
+            continue
+        # dict mapping lang to a set of expressions for a specific ingredient
+        seen: dict[str, set[str]] = defaultdict(set)
+        for field in ("names", "synonyms"):
+            for lang in ("xx", "en", "fr", "en", "es", "de", "nl", "it"):
+                names = getattr(node, field).get(lang)
+                if names is None:
+                    continue
+                if isinstance(names, str):
+                    # for 'names' field, the value is a string
+                    names = [names]
+
+                for name in names:
+                    normalized_name = fold(name.lower())
+                    if normalized_name in seen[lang] or len(normalized_name) <= 1:
+                        # Don't keep the normalized name if it already exists
+                        # or if it's length is <= 1 (to avoid false positives)
+                        # For example 'd' is a synonym for 'vitamin d'
+                        continue
+                    synonyms[(node.id, lang)].add(normalized_name)
+                    name_map[normalized_name].add((node.id, lang))
+                    seen[lang].add(normalized_name)
+
+    if add_synonym_combinations:
+        for normalized_name, keys in list(name_map.items()):
+            # get tokens from name by performing a simple whitespace
+            # tokenization
+            tokens = normalized_name.split(" ")
+            if len(tokens) <= 1:
+                continue
+
+            for current_node_id, current_lang in keys:
+                # combinations is a list of set of string, each element being
+                # a token with multiple possible synonyms.
+                # we initialize with the original tokens (if no synonym is
+                # found for any token, we will just generate the original
+                # normalized name again)
+                combinations = [{token} for token in tokens]
+                for token_idx in range(len(tokens)):
+                    token = tokens[token_idx]
+                    if token in name_map:
+                        # Lookup all ingredient IDs with the same lang that
+                        # match the normalized token string
+                        for key in (
+                            (node_id, lang)
+                            for (node_id, lang) in name_map[token]
+                            if lang == current_lang
+                        ):
```

Comment:
I find a bit difficult to read iterators in for loops !
```suggestion
                      lang_entries = (
                          (node_id, lang)
                          for (node_id, lang) in name_map[token]
                          if lang == current_lang
                      )
                      for key in lang_entries:
```